File: backend/internal/repository/mahasiswa_repository.go
package repository

import (
	"context"
	"database/sql"
	"fmt"

	"github.com/jmoiron/sqlx"

	"topsus-krs/internal/domain"
	"topsus-krs/pkg/pagination"
)

// MahasiswaRepository mendefinisikan kontrak operasi database untuk entitas Mahasiswa
type MahasiswaRepository interface {
	Create(ctx context.Context, m *domain.Mahasiswa) error
	FindAll(ctx context.Context, params *pagination.Params) ([]*domain.Mahasiswa, int64, error)
	FindByID(ctx context.Context, id uint64) (*domain.Mahasiswa, error)
	FindByNIM(ctx context.Context, nim string) (*domain.Mahasiswa, error)
	FindByEmail(ctx context.Context, email string) (*domain.Mahasiswa, error)
	FindByUserID(ctx context.Context, userID uint64) (*domain.Mahasiswa, error)
	FindByDosenPA(ctx context.Context, dosenID uint64, params *pagination.Params) ([]*domain.Mahasiswa, int64, error)
	Update(ctx context.Context, m *domain.Mahasiswa) error
	Delete(ctx context.Context, id uint64) error
	CountByJurusan(ctx context.Context) ([]map[string]interface{}, error)
	CountTotal(ctx context.Context) (int64, error)
}

type mahasiswaRepository struct {
	db *sqlx.DB
}

func NewMahasiswaRepository(db *sqlx.DB) MahasiswaRepository {
	return &mahasiswaRepository{db: db}
}

// baseSelectQuery adalah SELECT dengan JOIN ke tabel dosen untuk mendapatkan nama dosen PA
const baseSelectQuery = `
	SELECT
		m.id, m.user_id, m.nim, m.nama, m.email, m.jurusan,
		m.semester, m.dosen_pa_id, m.created_at, m.updated_at,
		COALESCE(d.nama, '') AS dosen_pa_nama
	FROM mahasiswa m
	LEFT JOIN dosen d ON m.dosen_pa_id = d.id
`

func (r *mahasiswaRepository) Create(ctx context.Context, m *domain.Mahasiswa) error {
	query := `
		INSERT INTO mahasiswa (user_id, nim, nama, email, jurusan, semester, dosen_pa_id)
		VALUES (:user_id, :nim, :nama, :email, :jurusan, :semester, :dosen_pa_id)
	`
	result, err := r.db.NamedExecContext(ctx, query, m)
	if err != nil {
		return err
	}
	id, err := result.LastInsertId()
	if err != nil {
		return err
	}
	m.ID = uint64(id)
	return nil
}

func (r *mahasiswaRepository) FindAll(ctx context.Context, params *pagination.Params) ([]*domain.Mahasiswa, int64, error) {
	var mahasiswas []*domain.Mahasiswa
	var total int64

	whereClause := ""
	args := []interface{}{}

	if params.Search != "" {
		whereClause = "WHERE (m.nama LIKE ? OR m.nim LIKE ? OR m.email LIKE ? OR m.jurusan LIKE ?)"
		searchTerm := fmt.Sprintf("%%%s%%", params.Search)
		args = append(args, searchTerm, searchTerm, searchTerm, searchTerm)
	}

	countQuery := fmt.Sprintf("SELECT COUNT(*) FROM mahasiswa m %s", whereClause)
	if err := r.db.GetContext(ctx, &total, countQuery, args...); err != nil {
		return nil, 0, err
	}

	query := fmt.Sprintf(
		"%s %s ORDER BY m.created_at DESC LIMIT ? OFFSET ?",
		baseSelectQuery, whereClause,
	)
	args = append(args, params.Limit, params.Offset)

	if err := r.db.SelectContext(ctx, &mahasiswas, query, args...); err != nil {
		return nil, 0, err
	}

	return mahasiswas, total, nil
}

func (r *mahasiswaRepository) FindByID(ctx context.Context, id uint64) (*domain.Mahasiswa, error) {
	var m domain.Mahasiswa
	query := fmt.Sprintf("%s WHERE m.id = ? LIMIT 1", baseSelectQuery)
	err := r.db.GetContext(ctx, &m, query, id)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	return &m, err
}

func (r *mahasiswaRepository) FindByUserID(ctx context.Context, userID uint64) (*domain.Mahasiswa, error) {
	var m domain.Mahasiswa
	query := fmt.Sprintf("%s WHERE m.user_id = ? LIMIT 1", baseSelectQuery)
	err := r.db.GetContext(ctx, &m, query, userID)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	return &m, err
}

func (r *mahasiswaRepository) FindByNIM(ctx context.Context, nim string) (*domain.Mahasiswa, error) {
	var m domain.Mahasiswa
	query := fmt.Sprintf("%s WHERE m.nim = ? LIMIT 1", baseSelectQuery)
	err := r.db.GetContext(ctx, &m, query, nim)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	return &m, err
}

func (r *mahasiswaRepository) FindByEmail(ctx context.Context, email string) (*domain.Mahasiswa, error) {
	var m domain.Mahasiswa
	query := fmt.Sprintf("%s WHERE m.email = ? LIMIT 1", baseSelectQuery)
	err := r.db.GetContext(ctx, &m, query, email)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	return &m, err
}

func (r *mahasiswaRepository) FindByDosenPA(ctx context.Context, dosenID uint64, params *pagination.Params) ([]*domain.Mahasiswa, int64, error) {
	var mahasiswas []*domain.Mahasiswa
	var total int64

	args := []interface{}{dosenID}
	searchClause := ""
	if params.Search != "" {
		searchClause = "AND (m.nama LIKE ? OR m.nim LIKE ? OR m.email LIKE ?)"
		searchTerm := fmt.Sprintf("%%%s%%", params.Search)
		args = append(args, searchTerm, searchTerm, searchTerm)
	}

	countQuery := fmt.Sprintf(
		"SELECT COUNT(*) FROM mahasiswa m WHERE m.dosen_pa_id = ? %s",
		searchClause,
	)
	if err := r.db.GetContext(ctx, &total, countQuery, args...); err != nil {
		return nil, 0, err
	}

	query := fmt.Sprintf(
		"%s WHERE m.dosen_pa_id = ? %s ORDER BY m.nama ASC LIMIT ? OFFSET ?",
		baseSelectQuery, searchClause,
	)
	args = append(args, params.Limit, params.Offset)

	if err := r.db.SelectContext(ctx, &mahasiswas, query, args...); err != nil {
		return nil, 0, err
	}

	return mahasiswas, total, nil
}

func (r *mahasiswaRepository) Update(ctx context.Context, m *domain.Mahasiswa) error {
	query := `
		UPDATE mahasiswa
		SET nim = :nim, nama = :nama, email = :email, jurusan = :jurusan,
		    semester = :semester, dosen_pa_id = :dosen_pa_id, updated_at = NOW()
		WHERE id = :id
	`
	_, err := r.db.NamedExecContext(ctx, query, m)
	return err
}

func (r *mahasiswaRepository) Delete(ctx context.Context, id uint64) error {
	_, err := r.db.ExecContext(ctx, `DELETE FROM mahasiswa WHERE id = ?`, id)
	return err
}

func (r *mahasiswaRepository) CountByJurusan(ctx context.Context) ([]map[string]interface{}, error) {
	rows, err := r.db.QueryContext(ctx, `
		SELECT jurusan, COUNT(*) as count
		FROM mahasiswa
		GROUP BY jurusan
		ORDER BY count DESC
	`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var result []map[string]interface{}
	for rows.Next() {
		var jurusan string
		var count int64
		if err := rows.Scan(&jurusan, &count); err != nil {
			return nil, err
		}
		result = append(result, map[string]interface{}{
			"jurusan": jurusan,
			"count":   count,
		})
	}
	return result, nil
}

func (r *mahasiswaRepository) CountTotal(ctx context.Context) (int64, error) {
	var total int64
	err := r.db.GetContext(ctx, &total, `SELECT COUNT(*) FROM mahasiswa`)
	return total, err
}

File: backend-refactor-python/app/repositories/mahasiswa_repository.py
from typing import Optional

from sqlalchemy import text
from sqlalchemy.orm import Session

_BASE = """
    SELECT
        m.id, m.user_id, m.nim, m.nama, m.email, m.jurusan,
        m.semester, m.dosen_pa_id, m.created_at, m.updated_at,
        COALESCE(d.nama, '') AS dosen_pa_nama
    FROM mahasiswa m
    LEFT JOIN dosen d ON m.dosen_pa_id = d.id
"""


def _row_to_dict(row) -> dict:
    return {
        "id":           row.id,
        "user_id":      row.user_id,
        "nim":          row.nim,
        "nama":         row.nama,
        "email":        row.email,
        "jurusan":      row.jurusan,
        "semester":     row.semester,
        "dosen_pa_id":  row.dosen_pa_id,
        "dosen_pa_nama": row.dosen_pa_nama or None,
        "created_at":   row.created_at,
        "updated_at":   row.updated_at,
    }


class MahasiswaRepository:
    def __init__(self, db: Session):
        self.db = db

    def find_all(self, page: int, limit: int, search: str = "") -> tuple[list[dict], int]:
        offset = (page - 1) * limit
        if search:
            like = f"%{search}%"
            where = "WHERE (m.nama LIKE :s OR m.nim LIKE :s OR m.email LIKE :s OR m.jurusan LIKE :s)"
            total = self.db.execute(
                text(f"SELECT COUNT(*) FROM mahasiswa m {where}"),
                {"s": like},
            ).scalar()
            rows = self.db.execute(
                text(f"{_BASE} {where} ORDER BY m.created_at DESC LIMIT :lim OFFSET :off"),
                {"s": like, "lim": limit, "off": offset},
            ).fetchall()
        else:
            total = self.db.execute(text("SELECT COUNT(*) FROM mahasiswa m")).scalar()
            rows = self.db.execute(
                text(f"{_BASE} ORDER BY m.created_at DESC LIMIT :lim OFFSET :off"),
                {"lim": limit, "off": offset},
            ).fetchall()
        return [_row_to_dict(r) for r in rows], int(total or 0)

    def find_by_id(self, id: int) -> Optional[dict]:
        row = self.db.execute(
            text(f"{_BASE} WHERE m.id = :id LIMIT 1"),
            {"id": id},
        ).fetchone()
        return _row_to_dict(row) if row else None

    def find_by_user_id(self, user_id: int) -> Optional[dict]:
        row = self.db.execute(
            text(f"{_BASE} WHERE m.user_id = :uid LIMIT 1"),
            {"uid": user_id},
        ).fetchone()
        return _row_to_dict(row) if row else None

    def find_by_nim(self, nim: str) -> Optional[dict]:
        row = self.db.execute(
            text(f"{_BASE} WHERE m.nim = :nim LIMIT 1"),
            {"nim": nim},
        ).fetchone()
        return _row_to_dict(row) if row else None

    def find_by_email(self, email: str) -> Optional[dict]:
        row = self.db.execute(
            text(f"{_BASE} WHERE m.email = :email LIMIT 1"),
            {"email": email},
        ).fetchone()
        return _row_to_dict(row) if row else None

    def find_by_dosen_pa(self, dosen_id: int, page: int, limit: int, search: str = "") -> tuple[list[dict], int]:
        offset = (page - 1) * limit
        if search:
            like = f"%{search}%"
            extra = "AND (m.nama LIKE :s OR m.nim LIKE :s OR m.email LIKE :s)"
            total = self.db.execute(
                text(f"SELECT COUNT(*) FROM mahasiswa m WHERE m.dosen_pa_id = :did {extra}"),
                {"did": dosen_id, "s": like},
            ).scalar()
            rows = self.db.execute(
                text(f"{_BASE} WHERE m.dosen_pa_id = :did {extra} ORDER BY m.nama ASC LIMIT :lim OFFSET :off"),
                {"did": dosen_id, "s": like, "lim": limit, "off": offset},
            ).fetchall()
        else:
            total = self.db.execute(
                text("SELECT COUNT(*) FROM mahasiswa m WHERE m.dosen_pa_id = :did"),
                {"did": dosen_id},
            ).scalar()
            rows = self.db.execute(
                text(f"{_BASE} WHERE m.dosen_pa_id = :did ORDER BY m.nama ASC LIMIT :lim OFFSET :off"),
                {"did": dosen_id, "lim": limit, "off": offset},
            ).fetchall()
        return [_row_to_dict(r) for r in rows], int(total or 0)

    def create(self, data: dict) -> int:
        result = self.db.execute(
            text(
                "INSERT INTO mahasiswa (user_id, nim, nama, email, jurusan, semester, dosen_pa_id) "
                "VALUES (:user_id, :nim, :nama, :email, :jurusan, :semester, :dosen_pa_id)"
            ),
            data,
        )
        self.db.commit()
        return result.lastrowid

    def update(self, id: int, data: dict) -> None:
        data["id"] = id
        self.db.execute(
            text(
                "UPDATE mahasiswa SET nim=:nim, nama=:nama, email=:email, jurusan=:jurusan, "
                "semester=:semester, dosen_pa_id=:dosen_pa_id, updated_at=NOW() WHERE id=:id"
            ),
            data,
        )
        self.db.commit()

    def delete(self, id: int) -> None:
        self.db.execute(text("DELETE FROM mahasiswa WHERE id = :id"), {"id": id})
        self.db.commit()

    def count_total(self) -> int:
        return int(self.db.execute(text("SELECT COUNT(*) FROM mahasiswa")).scalar() or 0)

    def count_by_jurusan(self) -> list[dict]:
        rows = self.db.execute(
            text("SELECT jurusan, COUNT(*) as count FROM mahasiswa GROUP BY jurusan ORDER BY count DESC")
        ).fetchall()
        return [{"jurusan": r.jurusan, "count": r.count} for r in rows]
